File: HWs/HW3/user/main.py
import sys
import socket
import struct

from validation import *
FIREWALL_TABLE_COLMUNS_NUM = 9

# User space program that will communicate with the kernel module.
# The program get an arguments and work by them.
# The supported values are:
# •	show_rules
# •	load_rules <path_to_rules_file>
# •	show_log
# •	clear_log


# load firewall rules
# through writing
# to /sys/class/fw/rules/rules
def load_rules(path_to_rules_file):
    rules_file_fw = open(path_to_rules_file, "r")
    rules_lines = rules_file_fw.readlines()
    fw_rules_fd = open("/sys/class/fw/rules/rules", "w")
    for rule_line in rules_lines:
        validate_rules_file_line(rule_line)
        parsed_rule_line = get_parsed_rule_line(rule_line)
        fw_rules_fd.write(parsed_rule_line)
    fw_rules_fd.close()

def get_parsed_rule_line(rule_line):
    direction_mapping = {"in": "1", "out": "2", "any": "3"}
    protocol_mapping = {"ICMP": "1", "TCP": "6", "UDP": "17", "any": "143"}
    ack_mapping = {"no": "0", "yes": "1", "any": "2"}
    action_mapping = {"drop": "0", "accept": "1"}

    # get fields from rule line
    rule_line_tokens = rule_line.split()
    parsed_rule_line_tokens = []

    rule_name = rule_line_tokens[0]
    parsed_rule_line_tokens.append(rule_name)
    
    direction = rule_line_tokens[1]
    parsed_rule_line_tokens.append(direction_mapping[direction])
    
    source_ip,  source_network_mask = rule_line_tokens[2].split("/")
    parsed_rule_line_tokens.append(ip_into_be_num(source_ip))
    parsed_rule_line_tokens.append(source_network_mask)

    dest_ip, dest_network_mask = rule_line_tokens[3].split("/")
    parsed_rule_line_tokens.append(ip_into_be_num(dest_ip))
    parsed_rule_line_tokens.append(dest_network_mask)

    protocol = rule_line_tokens[4]
    parsed_rule_line_tokens.append(protocol_mapping[protocol])

    source_port = rule_line_tokens[5]
    source_port_parsed = source_port if source_port != ">1023" else "1023"
    parsed_rule_line_tokens.append(source_port_parsed)
    
    dest_port = rule_line_tokens[6]
    dest_port_parsed = dest_port if dest_port != ">1023" else "1023"
    parsed_rule_line_tokens.append(dest_port_parsed)
    
    ack = rule_line_tokens[7]
    parsed_rule_line_tokens.append(ack_mapping[ack])

    action = rule_line_tokens[8]
    parsed_rule_line_tokens.append(action_mapping[action])

    parsed_rule_line_tokens = list(map(lambda field_value: str(field_value), parsed_rule_line_tokens))
    return " ".join(parsed_rule_line_tokens)

def ip_into_be_num(ip):
    packedIP = socket.inet_aton(ip)
    return struct.unpack("<L", packedIP)[0]

# show firewall rules
# through reading
# from /sys/class/fw/rules/rules
def show_rules():
    show_fw_rules_fd = open("/sys/class/fw/rules/rules", "r")
    show_fw_rules_fd.read()
    show_fw_rules_fd.close()

# show firewall logs
# through reading
# from /dev/fw_log
def show_log():
    fw_logs_fd = open("/dev/fw_log", "r")
    fw_logs_fd.read()
    fw_logs_fd.close()

# clear firewall logs
# through writing anything (N/A)
# to /sys/class/fw/log/reset
def clear_log():
    fw_logs_reset_fd = open("/sys/class/fw/log/reset", "w")
    fw_logs_reset_fd.write("0")
    fw_logs_reset_fd.close()

    fw_logs_clear_fd = open("/dev/fw_log", "w")
    fw_logs_clear_fd.truncate()
    fw_logs_clear_fd.close()

def main(argc, argv):
    validate_user_input(argc, argv)
    perform_load = (argc == 3) and argv[1] == "load_rules"
    if perform_load:
        load_rules(argv[2])
    elif argv[1] == "show_rules":
        show_rules()
    elif argv[1] == "show_log":
        show_log()
    elif argv[1] == "clear_log":
        clear_log()


if __name__ == "__main__":
    main(len(sys.argv), sys.argv)

File: HWs/HW3/user/main.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <ctype.h>
#include "main.h"

#define RULES_ATTR_PATH "/sys/class/fw/rules/rules"
#define RESET_ATTR_PATH "/sys/class/fw/log/reset"
#define FW_LOG_DEVICE "/dev/fw_log"

// User space program that will communicate with the kernel module.
// The program get an arguments and work by them.
// The supported values are:
// •	show_rules
// •	load_rules <path_to_rules_file>
// •	show_log
// •	clear_log

// load firewall rules
// through writing
// to /sys/class/fw/rules/rules
void load_rules(char *path_to_rules_file)
{
    FILE *firewall_new_rules_file_fp;
    char rule_chars_line[500] = {0};
    rule_t rule;
    int firewall_update_rules_fp;
    int i = 0;

    printf("User space load rules\n");

    firewall_new_rules_file_fp = fopen(path_to_rules_file, "r");
    firewall_update_rules_fp = open(RULES_ATTR_PATH, O_WRONLY);

    while(fgets(rule_chars_line, 500, firewall_new_rules_file_fp)) {
        validate_rules_file_line(rule_chars_line);
        parse_line_to_rule(&rule, rule_chars_line);
        print_rule(&rule);
        // printf("print user space chars of the rule load rules:\n");
        // for(i = 0; i < 60; i++) {
        //     printf("%i-%c-%d\t", i, ((char *)&rule)[i], ((char *)&rule)[i]);
        // }
        write(firewall_update_rules_fp, &rule, sizeof(rule));
    }

    fclose(firewall_new_rules_file_fp);
    close(firewall_update_rules_fp);
}

// show firewall rules
// through reading
// from /sys/class/fw/rules/rules
void show_rules(void)
{
    int i;
    int show_fw_rules_fp;
    // TODO: fix this to the real size of the rule structure from fw.h
    unsigned int RULE_SIZE = 100;
    // TODO: fix this to the real log structure from fw.h
    // rule_t static_rules_table[MAX_RULES];
    rule_t rule;
    char *char_ptr_rule = (char *) &rule;

    printf("User space show rules\n");

    show_fw_rules_fp = open(RULES_ATTR_PATH, O_RDONLY);

    read(show_fw_rules_fp, char_ptr_rule, RULE_SIZE);

    print_rule_in_format(&rule);

    close(show_fw_rules_fp);
}

// show firewall logs
// through reading
// from /dev/fw_log
void show_log(void)
{
    int i;
    int fw_logs_fp;
    // TODO: fix this to the real size of the log structure from fw.h
    unsigned int PACKET_LOG_SIZE = 100;
    // TODO: fix this to the real log structure from fw.h
    char log_buffer[100];

    printf("User space show log\n");

    fw_logs_fp = open(FW_LOG_DEVICE, O_RDONLY);

    // TODO: add a loop for reading and printing all the logs
    while (0)
    {
        read(fw_logs_fp, log_buffer, PACKET_LOG_SIZE);
    }
    // TODO: parse the <log_buffer> into the fields and print the fields
    printf("packet log\n");
    close(fw_logs_fp);
}

// clear firewall logs
// through writing anything (N/A)
// to /sys/class/fw/log/reset
void clear_log(void)
{
    int fw_logs_reset_fp;
    int fw_logs_clear_fp;

    printf("User space clear log\n");

    // deallocate the log resources
    fw_logs_reset_fp = open(RESET_ATTR_PATH, O_WRONLY);
    write(fw_logs_reset_fp, "0", strlen("0"));
    close(fw_logs_reset_fp);

    // clear the file from the logs using the write flag
    fw_logs_clear_fp = open(FW_LOG_DEVICE, O_WRONLY);
    write(fw_logs_clear_fp, "0", strlen("0"));
    close(fw_logs_clear_fp);
}

int main(int argc, char *argv[])
{
    validate_user_input(argc, argv);

    if (!strncmp(argv[1], "load_rules", strlen("load_rules")))
    {
        load_rules(argv[2]);
    }
    else if (!strncmp(argv[1], "show_rules", strlen("show_rules")))
    {
        show_rules();
    }
    else if (!strncmp(argv[1], "show_log", strlen("show_log")))
    {
        show_log();
    }
    else if (!strncmp(argv[1], "clear_log", strlen("clear_log")))
    {
        clear_log();
    }
}